### PR TITLE
ISSUE-133 Add support for most common IMAP commands

### PR DESCRIPTION
**Pull request by chibenwa**
- [x] CLOSE
 - [x] LOGOUT
 - [x] UNSELECT
 - [x] STATUS
 - [x] GETQUOTAROOT
 - [x] GETACL
 - [x] MYRIGHTS
 - [x] IDLE
 - [x] NOOP
 - [x] CHECK
 - [x] CAPABILITY
 - [x] ENABLE
 - [x] LSUB
 - [x] Namespace
 - [x] Search
 - [x] Uid search

This DSL could then serve as a base to write more complex IMAP scenarii.

Wish list to be feature complete on IMAP:

 - [ ] Subscribe
 - [ ] Unsubscribe
 - [ ] Create
 - [ ] Delete
 - [ ] Rename
 - [ ] Examine
 - [ ] Move
 - [ ] Copy
 - [ ] Uid expunge
 - [ ] Uid Store
 - [ ] Uid Move
 - [ ] Uid copy
 - [ ] Get quota
 - [ ] StartTLS
 - [ ] (compress)

### CODE REVIEW

**Comment by quantranhong1999** [comment]
I think I will implement other wish list commands in another PR to avoid potential blockers.
Also, is IMAP platform validation simulation modification missing?

**Comment by chibenwa** [comment]
> I think I will implement other wish list commands in another PR to avoid potential blockers.

Yes that's exactly the idea!

>  Also, is IMAP platform validation simulation modification missing? (Or you wait to add the wish list command)

1. it is in james gatling
2. and having a gatling imap dsl is different than running it

It will come in another jmes gatling PR

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessions.scala [hunk 1]
```diff
@@ -93,6 +93,45 @@ private class ImapSession(client: => ImapAsyncClient, protocol: ImapProtocol) ex
...
+      val handler = context.actorOf(IdleHandler.props(session), genName("idle"))
+      handler forward cmd
+    case cmd@Command.Logout(_) =>
+      val handler = context.actorOf(LogoutHandler.props(session), genName("logout"))
+      handler forward cmd
+    case cmd@Command.Lsub(_, _, _) =>
+      val handler = context.actorOf(LsubHandler.props(session), genName("lsub"))
+      handler forward cmd
+    case cmd@Command.MyRights(_, _) =>
+      val handler = context.actorOf(LsubHandler.props(session), genName("myrights"))
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/ImapSessions.scala:124]:
```suggestion
      val handler = context.actorOf(MyRightsHandler.props(session), genName("myrights"))
```

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetAclCommand.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.yahoo.imapnio.async.request.{ImapCommandType, ImapRequestAdapter}
+import io.netty.buffer.{ByteBuf, Unpooled}
+
+case object GetAclCommandType extends ImapCommandType {
+  override def getType: String = "GETACL"
+}
+
+class GetAclCommand(mailbox: String) extends ImapRequestAdapter {
+  override def getCommandLineBytes: ByteBuf = {
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetAclCommand.scala:13]: You mixed up code with `GetQuotaRootCommandType`

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetAclCommand.scala [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    val sb = Unpooled.buffer(GetQuotaRootCommandType.getType.getBytes(StandardCharsets.US_ASCII).length + mailbox.getBytes(StandardCharsets.US_ASCII).length + 100)
+    sb.writeBytes(GetQuotaRootCommandType.getType.getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes(" ".getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes(mailbox.getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes("\r\n".getBytes(StandardCharsets.US_ASCII))
+    sb
+  }
+
+  override def getCommandType: ImapCommandType = GetQuotaRootCommandType
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetAclCommand.scala:22]: Idem

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetAclHandler.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+      val responseCallback: Consumer[ImapAsyncResponse] = responses => {
+        import collection.JavaConverters._
+
+        val responsesList = ImapResponses(responses.getResponseLines.asScala.to[Seq])
+        logger.trace(s"On response for $userId :\n ${responsesList.mkString("\n")}")
+        self !  Response.AclResponse(responsesList)}
+      val errorCallback: Consumer[Exception] = _ => {}
+
+      val future = session.execute(new MyRightsCommand(mailbox))
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/GetAclHandler.scala:30]:
```suggestion
      val future = session.execute(new GetAclCommand(mailbox))
```
?

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/MyRightsCommand.scala [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.yahoo.imapnio.async.request.{ImapCommandType, ImapRequestAdapter}
+import io.netty.buffer.{ByteBuf, Unpooled}
+
+case object MyRightsCommandType extends ImapCommandType {
+  override def getType: String = "MYRIGHTS"
+}
+
+class MyRightsCommand(mailbox: String) extends ImapRequestAdapter {
+  override def getCommandLineBytes: ByteBuf = {
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/MyRightsCommand.scala:13]: Code mixed up with `GetQuotaRootCommandType`

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/MyRightsCommand.scala [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    val sb = Unpooled.buffer(GetQuotaRootCommandType.getType.getBytes(StandardCharsets.US_ASCII).length + mailbox.getBytes(StandardCharsets.US_ASCII).length + 100)
+    sb.writeBytes(GetQuotaRootCommandType.getType.getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes(" ".getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes(mailbox.getBytes(StandardCharsets.US_ASCII))
+    sb.writeBytes("\r\n".getBytes(StandardCharsets.US_ASCII))
+    sb
+  }
+
+  override def getCommandType: ImapCommandType = GetQuotaRootCommandType
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/MyRightsCommand.scala:22]: idem

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/MyRightsHandler.scala [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+import scala.collection.immutable.Seq
+
+object MyRightsHandler {
+  def props(session: ImapAsyncSession) = Props(new MyRightsHandler(session))
+}
+
+class MyRightsHandler(session: ImapAsyncSession) extends BaseActor {
+  override def receive: Receive = {
+    case Command.GetAcl(userId, mailbox) =>
```
Comment (Arsnael) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/MyRightsHandler.scala:19]: Command.MyRights ?

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

File: src/main/scala-2.12/com/linagora/gatling/imap/action/NamespaceAction.scala [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+class NamespaceAction(val imapContext: ImapActionContext,
+                      val requestName: String,
+                      override val checks: Seq[ImapCheck]) extends ValidatedActionActor with ImapActionActor {
+
+  override protected def executeOrFail(session: Session): Validation[_] = {
+    Validation.unit
+      .map(_ => {
+        val id: Long = session.userId
+        val handler = handleResponse(session, imapContext.clock.nowMillis)
+        sessions.tell(Command.Noop(UserId(id)), handler)
```
Comment (quantranhong1999) [inline src/main/scala-2.12/com/linagora/gatling/imap/action/NamespaceAction.scala:26]: I missed this. Should be Command.namespace?

Comment (chibenwa) [reply src/main/scala-2.12/com/linagora/gatling/imap/action/NamespaceAction.scala:26]: Sure!

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/SearchHandler.scala [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+object SearchHandler {
+  def props(session: ImapAsyncSession) = Props(new SearchHandler(session))
+}
+
+class SearchHandler(session: ImapAsyncSession) extends BaseActor {
+  override def receive: Receive = {
+    case Command.Search(userId, range, term) =>
+      logger.trace(s"LoginHandler for user : ${userId.value}, on actor ${self.path} responding to ${sender.path}")
+      context.become(waitForLoggedIn(sender()))
```
Comment (quantranhong1999) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/SearchHandler.scala:22]: Should be SearchHandler here and elsewhere in this file.

**Comment by quantranhong1999** [comment]
I started some wish list IMAP commands work here: https://github.com/linagora/gatling-imap/pull/63

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/UIDSearchHandler.scala [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+import scala.collection.immutable.Seq
+
+object UIDSearchHandler {
+  def props(session: ImapAsyncSession) = Props(new UIDSearchHandler(session))
+}
+
+class UIDSearchHandler(session: ImapAsyncSession) extends BaseActor {
+  override def receive: Receive = {
+    case Command.Search(userId, range, term) =>
```
Comment (quantranhong1999) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/UIDSearchHandler.scala:20]:
```suggestion
    case Command.UIDSearch(userId, range, term) =>
```

File: src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/UIDSearchHandler.scala [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        logger.error("UidSearchHandler command failed", e)
+      }
+
+      val future = session.execute(new UidSearchCommand(range.asImap, term, null))
+      future.setDoneCallback(responseCallback)
+      future.setExceptionCallback(errorCallback)
+  }
+
+  def waitForLoggedIn(sender: ActorRef): Receive = {
+    case msg@Response.SearchResult(_) =>
```
Comment (quantranhong1999) [inline src/main/scala-2.12/com/linagora/gatling/imap/protocol/command/UIDSearchHandler.scala:40]:
```suggestion
    case msg@Response.UIDSearchResult(_) =>
```

File: src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapSearchScenario.scala [hunk 12]
```diff
@@ -36,6 +38,7 @@ object ImapSearchScenario {
       .exec(imap("select").select("INBOX").check(ok))
       .exec(populateInbox)
       .pause(1 second)
-      .exec(imap("search").search(MessageRanges(Range(1, numberOfMailInInbox)), null))
+      .exec(imap("search").search(MessageRanges(Range(1, numberOfMailInInbox)), new FlagTerm(new Flags(Flags.Flag.SEEN), true)).check(ok))
+      .exec(imap("search").search(MessageRanges(Range(1, numberOfMailInInbox)), new FlagTerm(new Flags(Flags.Flag.SEEN), true)).check(ok))
```
Comment (quantranhong1999) [inline src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapSearchScenario.scala:42]: There should be a .uidSearch(...).

Comment (Arsnael) [reply src/main/scala-2.12/com/linagora/gatling/imap/scenario/ImapSearchScenario.scala]: Not solved

**Review by Arsnael — APPROVED** [review]

**Pull request merged**